File: go/31_next-permutation.go
func nextPermutation(nums []int) {
	var j int

	for i := len(nums) - 2; i >= 0; i-- {
		if nums[i] < nums[i+1] {
			j = i
			break
		}
		if i == 0 {
			j = -1
		}
	}

	if j != -1 {
		for i := len(nums) - 1; i > j; i-- {
			if nums[i] > nums[j] {
				nums[j], nums[i] = nums[i], nums[j]
				break
			}
		}
	}

	for i := 0; i < len(nums[(j+1):])/2; i++ {
		nums[j+i+1], nums[len(nums)-i-1] = nums[len(nums)-i-1], nums[j+i+1]
	}
}

File: python/31_next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        flag = False

        i = len(nums) - 2
        while i >= 0 and nums[i + 1] <= nums[i]:
            i -= 1
        if i == -1:
            i = 0
            flag = True

        if not flag:
            j = len(nums) - 1
            while j >= 0 and nums[j] <= nums[i]:
                j -= 1

            nums[i], nums[j] = nums[j], nums[i]
            i += 1

        j = len(nums) - 1
        while i <= j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1